hbt/production/tau.py

```diff
@@ -0,0 +1,317 @@
+# coding: utf-8
+
+"""
+Tau scale factor production.
+"""
+
+import functools
+import itertools
+
+from columnflow.production import Producer, producer
+from columnflow.util import maybe_import
+from columnflow.columnar_util import set_ak_column, flat_np_view, layout_ak_array
+
+ak = maybe_import("awkward")
+np = maybe_import("numpy")
+
+
+# helper
+set_ak_column_f32 = functools.partial(set_ak_column, value_type=np.float32)
+
+
+@producer(
+    uses={
+        "single_triggered", "cross_triggered",
```

Comment:
These columns are created by SelectEvents and then "horitontally" joined to the input array in ProduceColumns, right before the producer is called, so they should be present.

Reply:
OK, but how is the dependency resolved? It's not part of the requirements, and the selector isn't indicated anywhere, right?

Regardless how it works exactly, it's still working of course, which is good. But it might be nice to put a comment here to indicate where these columns come from

Reply:
They can actually come from different places, and I'd consider this a feature :)
E.g. calling the task that runs this producers would look like

```shell
law run cf.ProduceColumns \
    --version v1 \
    --dataset XYZ \
    --producer tau_weights \
    --selector my_hbt_selection
```

The selector has a default value, which is why one often doesn't use / see this, but the requirement, that e.g. the column `single_triggered` is existing, is set on the command line. In this case, `my_hbt_selection` is providing them and I think it's good that the tau weights producer is not hard wired to `my_hbt_selection` because it allows to run a different selection, say `my_other_selection` that would provide the same columns, perhaps with different content.

Reply:
Hm, I see. I do get your point about convenience. However, to me one of the strongest points in columnflow (and derived projects) is the clear dependency and a clear understanding where something is coming from when reading the code. Doing it like this is of course very flexible, but it also create a degeneracy in the sense that the reader (e.g. someone new to the project who wants to contribute) can't really understand where e.g. `single_triggered` is coming from...

I'm not saying we should disregard this kind of syntax - as you clearly pointed out, there is a very useful usecase here. But I would strongly be in favor of at least hinting where something is coming from, or at least what kind of object (e.g. `Producer`, `Selector`, ...) can produce these columns. I feel like this would greatly improve the readability of the code. However, I also don't want to force my personal opinion and preferences on everyone :D If you disagree, that's also fine and I'll resolve the comment, but I'd like to ask you to consider this issue

Reply:
I'm with you on the readability and documentation argument. Maybe we should always separate the "used" columns into at least two blocks and describe where things come from, e.g.

```python
@producer(
    uses={
        # custom columns, created by a selector
        "single_triggered", "cross_triggered",
        # columns from nano input files
        "nTau", "Tau.pt", "Tau.eta", "Tau.genPartFlav", "Tau.decayMode",
    },
    ...
```

I think this could very much help.

Reply:
Added that in the last commit 👍

Reply:
This is great, thanks! I'm now very happy :) And I completely agree with your proposal to always "separate" the uses columns by origin :+1: